doc/design/polymorphic_handler.md

```diff
@@ -0,0 +1,242 @@
+# Polymorphic Handler
+
+## Goals
+
+1. Manage a singleton handler (the current handler) that inherits from some interface `I`
+1. Initialize the handler with some default instance of a class that inherits from `I`
+1. Replace the singleton handler at runtime with a singleton of a different class 
+that also inherits from `I`
+1. Ensure that there is some handler to access at all times (until program termination)
+1. Ensure that any accessed singleton handler lives long enough to be accessed by other static
+   variables
+1. Obtaining the current handler must be thread-safe
+1. It must be possible to finalize the handler, i.e. prohibit any changes after it is finalized.
+1. Any attempt to change the handler after it is finalized, shall call an function that has access
+   to the current and new handler (for e.g. logging).
+
+To achieve this, we define another support class `StaticLifeTimeGuard` that solves the singleton lifetime problem.
+This class can be used on its own and is based on the nifty counter reference counting.
+
+The handler itself may not be thread-safe and multiple threads can concurrently use it. 
```

Comment:
This is a bit misleading. Do you mean the instance itself may not be thread-safe?

What do you think about using the terminology `underlying type` or `inner type` to prevent ambiguity?